### PR TITLE
Port to psycopg3

### PR DESCRIPTION
**Pull request by j-bennet**
## Description

Since psycopg2 was split into 2 packages (source-only and binary-only package) starting from versions 2.8, we have not been able to provide `pgcli` users a simple way to choose whether they want to install one or the other, or provide a "default" and a "fallback" option. Now that [psycopg3](https://www.psycopg.org/psycopg3/docs/index.html) is available, we can switch to that, and hopefully solve this problem once and for all.

For more context, see discussion [here](https://github.com/dbcli/pgcli/issues/1085#issuecomment-1047289518).

~Because psycopg3 only works with Python 3.7 and up, we have to drop support for Python 3.6 here. Which is fine, because EOL for Python 3.6 happened 2 months ago (December 23, 2021). But this is a breaking change, and I should bump the major version to release this.~

## Checklist

- [x] I've added this contribution to the `changelog.rst`.
- [x] I installed pre-commit hooks (`pip install pre-commit && pre-commit install`), and ran `black` on my code.
- [x] Please squash merge this pull request (uncheck if you'd like us to merge as multiple commits)

### CODE REVIEW

**Comment by j-bennet** [comment]
@dvarrazzo I can't tag you to review, but I appreciate it if you could take a look.

File: pgspecial/dbcommands.py [hunk 1]
```diff
@@ -110,17 +105,17 @@ def list_roles(cur, pattern, verbose):
...
-        sql += "WHERE r.rolname ~ %s"
-        params.append(schema)
-    sql = cur.mogrify(sql + " ORDER BY 1", params)
+        sql += "WHERE r.rolname ~ %(rolname)s"
+        params["rolname"] = schema
+    sql += " ORDER BY 1"
 
-    log.debug(sql)
-    cur.execute(sql)
+    log.debug(f"{sql}, {params}")
```
Comment (lelit) [inline pgspecial/dbcommands.py:115]: Probably with an negligible impact in this context, but usually you don't pass f-string to log, but rather use the "lazy" good-n-old printf-like idiom:
```
$ python3 -m timeit -s "import logging; x = ['foo']*100" "logging.debug(f'x = {x}')"
50000 loops, best of 5: 6 usec per loop
$ python3 -m timeit -s "import logging; x = ['foo']*100" "logging.debug('x = %s', x)"
500000 loops, best of 5: 583 nsec per loop

```

**Comment by dvarrazzo** [comment]
> Because psycopg3 only works with Python 3.7 and up, 

Hi, @j-bennet, 

Psycopg 3.0.x is still compatible with Python 3.6. only in the main branch, to be released as 3.1, we have dropped 3.6 support, see [changelog](https://www.psycopg.org/psycopg3/docs/news.html).

I think you should be able to set a dependency of psycopg >= 3.0.10 for pgcli, which will work for python 3.6 too. That's still unreleased, but it should by the time the branch is ready. If you need a dev requirement, from the moment you can use the [fix-231 branch](https://github.com/psycopg/psycopg/tree/fix-231) for reasons already discussed, also working on python 3.6.

File: pgspecial/dbcommands.py (hunk 1 above)
Comment (j-bennet) [reply pgspecial/dbcommands.py:115]: 👍  good point.

File: tests/dbutils.py [hunk 2]
```diff
@@ -31,8 +30,9 @@ def db_connection(dbname=None):
 try:
     conn = db_connection(dbname=None)
     CAN_CONNECT_TO_DB = True
-    SERVER_VERSION = conn.server_version
-except:
+    SERVER_VERSION = conn.info.server_version
+except Exception as x:
+    print(x)
```
Comment (amjith) [inline tests/dbutils.py:35]: What's this print used for?

Comment (amjith) [reply tests/dbutils.py:35]: Didn't realize it was a test file. :+1:

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@ jobs:
 
     strategy:
       matrix:
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: ['3.7', '3.8', '3.9', '3.10']
```
Comment (dvarrazzo) [inline .github/workflows/ci.yml:14]: As mentioned in the ticket, no need to drop by 3.6 in order to depend on psycopg 3.0.x

**Review by dvarrazzo — COMMENTED** [review]
Left a few comments, but looks good :)

File: tests/dbutils.py (hunk 2 above)
Comment (j-bennet) [reply tests/dbutils.py:35]: 🤦

File: pgspecial/iocommands.py [hunk 4]
```diff
@@ -254,9 +247,8 @@ def execute_named_query(cur, pattern, **_):
             if query is None:
                 raise Exception("Bad arguments\n" + params)
         cur.execute(query)
-    # pg3: (but psycopg2 too): you can use `except psycopg.errors.SyntaxError`
-    except psycopg2.ProgrammingError as e:
-        if e.pgcode == psycopg2.errorcodes.SYNTAX_ERROR and "%s" in query:
+    except psycopg.errors.SyntaxError:
+        if "%s" in query:
```
[the remark this answers is not in the document]
Comment (chagui) [reply pgspecial/iocommands.py:251]: The feature is part of Named Queries and documented here: https://www.pgcli.com/named_queries.md

File: pgspecial/iocommands.py [hunk 5]
```diff
@@ -148,41 +137,45 @@ def copy(cur, pattern, verbose):
...
+                if not data:
+                    break
+                pgcopy.write(data)
     else:
-        return [(None, None, None, cur.statusmessage)]
+        with cur.copy("copy " + query) as pgcopy:
+            for data in pgcopy:
+                file.write(bytes(data))
+
+    # TODO: pg3 raises an exception here, is this a bug?
```
[the remark this answers is not in the document]
Comment (j-bennet) [reply pgspecial/iocommands.py:171]: Here is a minimal reproduction case:

```
import psycopg

conn = psycopg.connect(
    user='postgres',
    host='127.0.0.1',
    password='',
    port=5432,
    dbname=None
)
conn.autocommit = True
with conn.cursor() as cur:
    with cur.copy("COPY (select 'This', 'Is', 'Text') TO STDOUT") as copy:
        for row in copy.rows():
            print(row)
    print('---')
    print(cur.description)
    print('---')
```

The exception is thrown on line `print(cur.description)`:

```
Traceback (most recent call last):
  File "/Users/irinatruong/Library/Application Support/JetBrains/PyCharm2020.2/scratches/scratch.py", line 16, in <module>
    print(cur.description)
  File "/Users/irinatruong/.pyenv/versions/pgspecial/lib/python3.9/site-packages/psycopg/cursor.py", line 127, in description
    return [Column(self, i) for i in range(res.nfields)]
  File "/Users/irinatruong/.pyenv/versions/pgspecial/lib/python3.9/site-packages/psycopg/cursor.py", line 127, in <listcomp>
    return [Column(self, i) for i in range(res.nfields)]
  File "/Users/irinatruong/.pyenv/versions/pgspecial/lib/python3.9/site-packages/psycopg/_column.py", line 32, in __init__
    raise e.InterfaceError(f"no name available for column {index}")
psycopg.InterfaceError: no name available for column 0
```

[the remark this answers is not in the document]
Comment (dvarrazzo) [reply pgspecial/iocommands.py:171]: Great, thank you! The result object received from copy is incomplete in unexpected ways it seems. Will fix it in 3.0.10

**Comment by dvarrazzo** [comment]
FYI, we have released psycopg 3.0.10, which fixes the Ctrl-C interaction and the error reported in the column names after copy.

File: setup.py [hunk 6]
```diff
@@ -25,7 +25,7 @@
     install_requires=[
         "click >= 4.1",
         "sqlparse >= 0.1.19",
-        "psycopg2 >= 2.7.4",
+        "psycopg >= 3.0.9,<3.1.0",
```
Comment (dvarrazzo) [inline setup.py:28]: should change this to >= 3.0.10. I advice to drop the <3.1.0: there is now abundant literature advising against using an upper bound in dependencies.

File: requirements-dev.txt [hunk 7]
```diff
@@ -1,6 +1,6 @@
 pytest>=6.2.4
 tox>=1.9.2
-psycopg2>=2.5.4
+psycopg>=3.0.9,<3.1.0
```
Comment (dvarrazzo) [inline requirements-dev.txt:3]: Change consistently with setup.py

**Comment by SuperSandro2000** [comment]
I don't understand why this is marked as a breaking change in the changelog. I did some limited testing and so far everything just worked. There is also no mention of what exactly is the breaking change and how to update or I didn't find it.

**Pull request merged**